### ISSUE TITLE
Fix integration test "FreeText Editor FreeText on several pages must check that first annotation is selected without errors"

### ISSUE THREAD

**Issue by timvandermeij** [mozilla/pdf.js#17087]
Introduction date: September 7, 2023
Introduction PR: https://github.com/mozilla/pdf.js/commit/8ab4e2e6e7aa39288d502b16016de2208be10d67 (/cc @calixteman)
Type: intermittent (sometimes it works, indicating a possible timing/waiting issue)
OS: Windows and Linux
Logs: http://54.193.163.58:8877/972b6fbf01d54b8/output.txt (Windows) and http://54.241.84.105:8877/6ce0d1567eb597b/output.txt (Linux)
Traceback:

```
Failures:
1) FreeText Editor FreeText on several pages must check that first annotation is selected without errors
  Message:
[31m    Error: Timeout - Async function did not complete within 30000ms (set by jasmine.DEFAULT_TIMEOUT_INTERVAL)[0m
  Stack:
        at <Jasmine>
        at listOnTimeout (node:internal/timers:569:17)
        at process.processTimers (node:internal/timers:512:7)

<snip>

Suite error: Interaction in secHandler.pdf
  Message:
[31m    ProtocolError: Waiting for selector `.page[data-page-number = "14"] > .annotationEditorLayer.freetextEditing` failed: Runtime.callFunctionOn timed out. Increase the 'protocolTimeout' setting in launch/connect calls for a higher timeout if needed.[0m
  Stack:
        at <instance_members_initializer> (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/Connection.js:39:14)
        at new Callback (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/Connection.js:43:16)
        at CallbackRegistry.create (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/Connection.js:82:26)
        at Connection._rawSend (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/Connection.js:203:26)
        at CdpCDPSession.send (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/CDPSession.js:58:33)
        at #evaluate (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/ExecutionContext.js:208:50)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async ExecutionContext.evaluateHandle (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/ExecutionContext.js:175:16)
        at async IsolatedWorld.evaluateHandle (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/IsolatedWorld.js:121:16)
        at async CdpJSHandle.evaluateHandle (file:///C:/pdfjs/botio-files-pdfjs/private/972b6fbf01d54b8/node_modules/puppeteer-core/lib/esm/puppeteer/api/JSHandle.js:163:20)
```

This one is especially annoying because timeouts occurring in the freetext editor spec leak through to other tests, causing them to fail. Apparently e.g. protocol errors are raised in later tests because the test runner has already moved on. Found cases of this:

- http://54.193.163.58:8877/972b6fbf01d54b8/output.txt (`secHandler` doesn't use the mentioned selector, but the test here does).
- http://54.193.163.58:8877/3c7a1b26b910079/output.txt (`issue14307` doesn't use the mentioned selector, but the test here does).
- http://54.241.84.105:8877/6ce0d1567eb597b/output.txt (the focus/blur test lives in the sprint spec, but the traceback for the protocol error points to the freetext editor spec).

**Comment by calixteman** [comment]
Unfortunately, I think the code itself is correct:
https://github.com/mozilla/pdf.js/blob/e6c325717570b9db2c8cf46342663b6c80c9cef7/test/integration/freetext_editor_spec.js#L2575-L2582
for any reason the pressed key is ignored or doesn't work as expected.
Maybe we should do something similar to:
https://github.com/mozilla/pdf.js/blob/e6c325717570b9db2c8cf46342663b6c80c9cef7/test/integration/stamp_editor_spec.js#L63
for any reason ctrl+v was not always working correctly hence the while and the timeout (500).

**Comment by calixteman** [comment]
I wonder if it could be caused by:
https://github.com/mozilla/pdf.js/blob/f2c9b64cb2001fd36ccabed09f32d43420f03b50/src/display/editor/annotation_editor_layer.js#L404-L419

**Comment by timvandermeij** [comment]
I'm not sure why, but it looks like if actions time out in particular in the freetext editor spec that the protocol errors caused by it cause subsequent tests to fail because they are raised when those run. That does explain why we have had quite a bit of seemingly random failures in unrelated tests lately; I have updated the list in the top comment to illustrate this for the cases where I know for sure the actual test was not the culprit of it failing.

I hope that if we figure out this one that that will end the intermittents for good (because all the previous work is really showing effects already in that the runs are much more stable now).

**Comment by timvandermeij** [comment]
For now I'm going to assume all intermittent failures related to the freetext editor spec share the same root cause because there is not really a pattern in them, so I'll only create issues for intermittents we spot from now on that are related to other specs or that are freetext-related but have a clear cause. I hope the PR above will help to shed more light on which freetext editor spec tests _actually_ fail as opposed to where the protocol error just happens to end up since I think that's blurring our vision here.

**Comment by timvandermeij** [comment]
Now that we fixed the other ones, this (and perhaps related FreeText) intermittent seems to be the only remaining one. I did some more debugging in #17126 but don't really have a much better idea yet about what fails here, so this probably requires more debugging. In the meantime I'll be monitoring the intermittents the coming week to see if it's indeed only this one that remains and if we can think of a possible root cause here.
